### PR TITLE
Powercell reagent effects framework

### PR DESCRIPTION
**Pull request by Time-Green**
### Adds a framework for easy adding of new reactions to powercells.

_But how do I make reactions happun?11!!1!_
Inject shit into a powercell and drain energy

Also includes a bananium reaction and an uranium reaction.
Both cancer inducing.

:cl:
rscadd: Adds a framework for powercell reactions
rscadd: Adds a bananium and uranium reaction for the above^
/:cl:
Also moves some file

### CODE REVIEW

File: code/__DEFINES/is_helpers.dm [hunk 1]
```diff
@@ -27,6 +27,10 @@
 #define isashwalker(A) (is_species(A, /datum/species/lizard/ashwalker))
 #define ischiefwalker(A) (is_species(A, /datum/species/lizard/ashwalker/chieftain))
 #define iskrampus(A) (is_species(A, /datum/species/demon))
+#define isplant(A) (is_species(A, /datum/species/plant))
+#define isphytosian(A) (is_species(A, /datum/species/plant))
+#define ispreternis(A) (is_species(A, /datum/species/android))
```
Comment (Time-Green) [inline code/__DEFINES/is_helpers.dm]: And this because it's really bothering me this doesn't exist.

**Comment by Cruix** [comment]
Can't you just put these procs in /datum/reagent?

**Comment by Time-Green** [comment]
It doesn't actually do much with reagents, it's more based on the powercell. I figured their purposes were to different for it to be a subtype

File: code/__DEFINES/is_helpers.dm (hunk 1 above)
Comment (ShadowDeath6) [inline code/__DEFINES/is_helpers.dm]: why the fuck do you have two defines for the same thing

Comment (ShadowDeath6) [reply code/__DEFINES/is_helpers.dm]: why do you have isplant and isphywhatever for the exact same thing, make it isplant or isplantperson or something

**Comment by ShadowDeath6** [comment]
If this makes borgs able to irradiate people then :+1:

File: code/__DEFINES/is_helpers.dm (hunk 1 above)
Comment (Time-Green) [reply code/__DEFINES/is_helpers.dm]: I wasnt sure wich one to use, and figured people might get confused so I made both

Comment (ShadowDeath6) [reply code/__DEFINES/is_helpers.dm]: They can look in ishelpers, that's what the file is for

Comment (Time-Green) [reply code/__DEFINES/is_helpers.dm]: I guess, I'll go and remove one when I get the chance

**Comment by Time-Green** [comment]
It does, but at a huge powercost and random stuns

**Comment by ShadowDeath6** [comment]
don't make it stun

**Comment by Time-Green** [comment]
It gets a short stun like every 10 seconds.
Dude, the rads do alot of damage.

**Comment by Time-Green** [comment]
Dammit, I just now see the merge conflict

**Comment by Time-Green** [comment]
Fixed the conflict

**Comment by ArgyleSweatshop** [comment]
This won't work at all. At least I don't think.

You've created a separate datum- powercell_reagents
And done nothing to associate them with actual reagents. Later, when you inject a battery, it checks if what you inject is a powercell_reagent, which it never can be.

**Comment by ArgyleSweatshop** [comment]
Maybe that's intentional, though, since it's just a framework.

**Comment by Time-Green** [comment]
I have tested every single bit, and it all works fine. 
I'll absolutely tinker with it once it gets merged, but this works. I tested it and it works. 
I tested the small objects I added, and it works.

Also, I cannot read what or where you commented on my code

**Comment by ShadowDeath6** [comment]
@Time-Green 
https://github.com/Time-Green/-tg-station-yog-rebase/commit/801a85a1ff3d7277f93e3244deaa3d0f47fe078e#diff-0eae04c47105ad93b26d765f35fbb175R161

File: code/modules/power/cell/powercell.dm [hunk 2]
```diff
@@ -1,309 +1,339 @@
...
+	else
+		user << "The charge meter reads [round(src.percent() )]%."
+
+/obj/item/weapon/stock_parts/cell/suicide_act(mob/user)
+	user.visible_message("<span class='suicide'>[user] is licking the electrodes of the [src.name]! It looks like \he's trying to commit suicide.</span>")
+	return (FIRELOSS)
+
+/obj/item/weapon/stock_parts/cell/attackby(obj/item/W, mob/user, params)
+	..()
+	world << "5 [W]"
```
Comment (X-TheDark) [inline code/modules/power/cell/powercell.dm]: ZA WARUDO called.

File: code/modules/power/cell/powercell.dm [hunk 3]
```diff
@@ -1,309 +1,339 @@
...
+	if(istype(W, /obj/item/weapon/reagent_containers/syringe))
+		world << "6 [W]"
```
Comment (X-TheDark) [inline code/modules/power/cell/powercell.dm]: And again...

File: code/modules/power/cell/powercell.dm [hunk 4]
```diff
@@ -1,309 +1,339 @@
...
+		var/obj/item/weapon/reagent_containers/syringe/S = W
+		user << "<span class='notice'>You inject the solution into the power cell.</span>"
+		world << "7 [S]"
```
Comment (X-TheDark) [inline code/modules/power/cell/powercell.dm]: And again...

File: code/modules/power/cell/powercell.dm [hunk 5]
```diff
@@ -1,309 +1,339 @@
...
+		for(var/A in specials)
+			world << "1 [A]"
```
Comment (X-TheDark) [inline code/modules/power/cell/powercell.dm]: You know the drill...

File: code/modules/power/cell/powercell.dm [hunk 6]
```diff
@@ -1,309 +1,339 @@
...
+			if(S.reagents.has_reagent(A, 5))
+				world << "3 [A]"
```
Comment (X-TheDark) [inline code/modules/power/cell/powercell.dm]: And obviously, you know the world too...

File: code/modules/power/cell/powercell.dm [hunk 7]
```diff
@@ -1,309 +1,339 @@
...
+				var/datum/powercell_reagents/R = specials[A]
+				special = new R
+				world << "4 [R]"
```
Comment (X-TheDark) [inline code/modules/power/cell/powercell.dm]: But you seem a bit too desperate...don't call them so often.

**Review by X-TheDark — CHANGES_REQUESTED** [review]
You left a little bit of something there.

File: code/modules/power/cell/powercell.dm (hunk 7 above)
Comment (Time-Green) [reply code/modules/power/cell/powercell.dm]: what the shit, I thought I removed those

**Review by X-TheDark — APPROVED** [review]
Debug code has been removed.

**Comment by Mrcreepssss** [comment]
the idea would be cool, but what sorts of things would it do? how would it be used? would it be usefull at all? will it not be broken?

**Comment by ShadowDeath6** [comment]
Still want this to not stun borgs

**Comment by Time-Green** [comment]
@Mrcreepssss 
How would it be used?
You inject a needle filled with a reagent, once charge gets removed from the cell, through whatever way, a special effect happens. 
Will it be useful at all? 
Yes, I managed to kill myself and dominus while messing around with it on live.
Will it not be broken?
Of course not, I coded it, what could possibly go wrong?

And shadow, do you have a better idea? I think them giving everyone cancer in a close proximity to them should have some heavy cost.

**Comment by ShadowDeath6** [comment]
Make it drain their cells when it pulses or something

**Comment by Time-Green** [comment]
I'll look at it this week, and that's actually already being done.

**Comment by ShadowDeath6** [comment]
then make it drain more idk

**Comment by Mrcreepssss** [comment]
the main post gives the impression the cell pulses randomly, but you said it happens when emptied so ehm okay that's still not very controllable (imagine putting one of these cells into a defib tho.) maybe there should be a way to trigger it early? perhaps have it trigger if removed without another thing being done? or when emagged?
and perhaps have the rads it emits deal some slight burn damage to the borg? it makes sense enough as there's suddenly a spike of radiation (which heats it up) from the cell.

and ehh, how does this work with slimecells? or other self-recharging cells?

**Pull request closed without merging**